### PR TITLE
issue#4 : Handle large msgs

### PR DESCRIPTION
**Pull request by ShikhaChauhan14**

### CODE REVIEW

File: cpp/src/server/handleMsg.cc [hunk 1]
```diff
@@ -4,20 +4,50 @@ void handleMsg(
...
+  // details have chatroom name and personal msg receiver's name
+  auto details = tokens[0];
+  string receiverName;
+  string chatRoomName;
+  //If msg is personal
+  auto index = details.find('@');
+  if(index != string::npos) {
+    receiverName = details.substr(index + 1, details.substr(index + 1).find(' '));
+    chatRoomName = details.substr(0, index); 
+  }
```
Comment (sidvishnoi) [inline cpp/src/server/handleMsg.cc:37]: Consistency:
``` cpp
} else {   // Not personal msg
```

File: cpp/src/client/clientChat.cc [hunk 2]
```diff
@@ -46,7 +46,19 @@ int clientChat(int sockfd, const std::string &username) {
...
+        auto index = msgToSend.find('@');
+        // send private message.
+        std::string receiverName;
+        if (index != std::string::npos) {
+          receiverName = msgToSend.substr(index + 1, msgToSend.substr(index + 1).find(' '));
+        }
+        std::string msg ;
+        if (receiverName != "") {
+          msg = chatroom + "@" + receiverName + DELIM + msgToSend;
+        }
```
Comment (sidvishnoi) [inline cpp/src/client/clientChat.cc:58]: ` } else { ` in same line

File: cpp/src/client/clientChat.cc [hunk 3]
```diff
@@ -59,6 +71,14 @@ int clientChat(int sockfd, const std::string &username) {
         }
         case cmd::JOIN: {
           auto chatroom = split(msgToSend, " ", 2)[1].substr(1);
+          //Validate chatroom name
+          auto index = chatroom.find('@');
+          if (index != std::string::npos) {
+            printMessage(
+              "ERROR" + DELIM + "SYNTAX" + DELIM +
+              "Cann't use @ symbol in chatroom name");
```
Comment (sidvishnoi) [inline cpp/src/client/clientChat.cc]: Typo: "Cann't"

File: cpp/src/client/clientChat.cc [hunk 4]
```diff
@@ -59,6 +71,14 @@ int clientChat(int sockfd, const std::string &username) {
         }
         case cmd::JOIN: {
           auto chatroom = split(msgToSend, " ", 2)[1].substr(1);
+          //Validate chatroom name
+          auto index = chatroom.find('@');
+          if (index != std::string::npos) {
```
Comment (sidvishnoi) [inline cpp/src/client/clientChat.cc:76]:
``` cpp
if (chatroom.find('@') != std::string::npos) {
```

File: cpp/src/client/printMessage.cc [hunk 5]
```diff
@@ -4,6 +4,10 @@
 void printMessage(const std::string &msg) {
   cout << "\r";
   auto tokens = split(msg, DELIM, 3);
+  if (tokens.size() != 3) {
```
Comment (sidvishnoi) [inline cpp/src/client/printMessage.cc:7]: why this?

File: cpp/src/server/handleMsg.cc [hunk 6]
```diff
@@ -4,20 +4,50 @@ void handleMsg(
   const int currentClientFd,
   const ChatroomToFdList &chatRooms,
   const FdToName &clients,
-  string msg
+  string msg,
+  bool &personalMsg,
```
Comment (sidvishnoi) [inline cpp/src/server/handleMsg.cc]: maybe pass this by value

File: cpp/src/server/handleMsg.cc [hunk 7]
```diff
@@ -4,20 +4,50 @@ void handleMsg(
...
+    }
+    return;
+  }
+  // details have chatroom name and personal msg receiver's name
+  auto details = tokens[0];
+  string receiverName;
+  string chatRoomName;
+  //If msg is personal
+  auto index = details.find('@');
+  if(index != string::npos) {
```
Comment (sidvishnoi) [inline cpp/src/server/handleMsg.cc:34]:
``` cpp
if(details.find('@') != string::npos) {
```

**Comment by sidvishnoi** [comment]
Can you tell the approach briefly? I think we can improve it

**Comment by ShikhaChauhan14** [comment]
Now client will send msgs like
chatroom + "@" + receiverName + DELIM + msgToSend 
if msg is private else in the previous format.

File: cpp/src/server/handleMsg.cc (hunk 6 above)
Comment (ShikhaChauhan14) [reply cpp/src/server/handleMsg.cc]: No

File: cpp/src/client/printMessage.cc (hunk 5 above)
Comment (ShikhaChauhan14) [reply cpp/src/client/printMessage.cc:7]: If token.size() != 3 it means received msg is the part of previously received msg, so need of formatting simply print the msg.

File: cpp/src/client/clientChat.cc (hunk 4 above)
Comment (ShikhaChauhan14) [reply cpp/src/client/clientChat.cc:76]: It means msg is not private.

File: cpp/src/client/clientChat.cc (hunk 3 above)
Comment (ShikhaChauhan14) [reply cpp/src/client/clientChat.cc]: Sorry

File: cpp/src/client/clientChat.cc (hunk 2 above)
Comment (ShikhaChauhan14) [reply cpp/src/client/clientChat.cc:58]: Sorry

File: cpp/src/client/printMessage.cc (hunk 5 above)
Comment (sidvishnoi) [reply cpp/src/client/printMessage.cc:7]: what if someone sends a normal msg (that needs to be formatted) while the larger message is in print/transition?

File: cpp/src/client/clientChat.cc (hunk 4 above)
Comment (sidvishnoi) [reply cpp/src/client/clientChat.cc:76]: yes got that. was saying `index` need not be assigned

File: cpp/src/server/handleMsg.cc (hunk 1 above)
Comment (ShikhaChauhan14) [reply cpp/src/server/handleMsg.cc:37]: Details having chatroom name and private msg receiver's name(If private msg)

Comment (sidvishnoi) [reply cpp/src/server/handleMsg.cc:37]: i think we can have a better variable name. One that has more explicit meaning

File: cpp/src/server/handleMsg.cc (hunk 7 above)
Comment (ShikhaChauhan14) [reply cpp/src/server/handleMsg.cc:34]: We are using index variable again in if block.

File: cpp/src/client/printMessage.cc (hunk 5 above)
Comment (sidvishnoi) [reply cpp/src/client/printMessage.cc:7]: add a comment here why it is necessary

File: cpp/src/server/serverChat.cc [hunk 8]
```diff
@@ -36,7 +37,8 @@ int serverChat(int sockfd) {
       if (!FD_ISSET(currentClientFd, &master)) continue;
 
       std::fill_n(buffer, BUFFER_SIZE, '\0');
-      const int bytesRead = recv(currentClientFd, buffer, sizeof(buffer), 0);
+      // sizeof(buffer) - 1 to avoid garbage values
+      const int bytesRead = recv(currentClientFd, buffer, sizeof(buffer)-1, 0);
```
Comment (sidvishnoi) [inline cpp/src/server/serverChat.cc:41]: no data is lost due to this 1 byte? please confirm

Comment (ShikhaChauhan14) [reply cpp/src/server/serverChat.cc:41]: No data loss, confirmed

**Pull request closed without merging**